### PR TITLE
Integrate catalog schema validation into planner -WIP

### PR DESCRIPTION
**Pull request by zachjsh**
### Description

This PR contains a portion of the changes from the inactive draft PR for integrating the catalog with the Calcite planner https://github.com/apache/druid/pull/13686 from @paul-rogers, allowing the datasource table schemas defined in the catalog to be validated against when ingesting data into the underlying datasource, during SQL based ingestion. This allows for SQL based ingestion into a table with a declared schema to produce segments that conform to that schema. If partitioning and clustering is not defined at ingestion time, defaults for these parameters, as defined in the catalog for the table, if found, are used.

TODO: add more tests.

#### Release note

<hr>

##### Key changed/added classes in this PR
 * `MyFoo`
 * `OurBar`
 * `TheirBaz`

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java [hunk 1]
```diff
@@ -19,38 +19,762 @@
...
+
+    // If this is a 2-part name, the first part must be the datasource schema.
+    if (n == 2 && !validatorContext.druidSchemaName().equals(destId.names.get(0))) {
+      throw new IAE("Cannot %s into [%s] because the table is not in the 'druid' schema",
+          operationName,
+          destId
+      );
+    }
+    try {
+      // Try to resolve the table. Will fail if this is an INSERT into a new table.
```
Comment (zachjsh) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:265]: This causes MSQ ingestion to fail on non-catalog tables.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java [hunk 2]
```diff
@@ -133,6 +141,41 @@
     this.context = config.getContext();
     this.connectionConfig = connConfig(context);
     this.typeSystem = config.getTypeSystem();
+
+    // With the catalog, schemas can provide functions.
+    // Add the necessary indirection. The type factory used here
+    // is the Druid one, since the per-query one is not yet available
+    // here. Nor are built-in function associated with per-query types.
+    this.operatorTable = new ChainedSqlOperatorTable(
```
[the remark this answers is not in the document]
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java:149]: Need to check on this.

File: sql/src/main/codegen/includes/common.ftl [hunk 3]
```diff
@@ -18,58 +18,51 @@
...
-  String unparseString;
+  SqlNode result;
 }
 {
   (
     <HOUR>
     {
-      granularity = Granularities.HOUR;
-      unparseString = "HOUR";
+      result = SqlLiteral.createCharString(DruidSqlParserUtils.HOUR_GRAIN, getPos());
```
Comment (kgyrtkirk) [inline sql/src/main/codegen/includes/common.ftl:30]: I wonder if it would be possible to use `SqlLiteral.createSymbol` here instead; that could remove the need for the string based matching as well...

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java [hunk 4]
```diff
@@ -133,6 +141,41 @@ public CalcitePlanner(FrameworkConfig config)
...
```
Comment (kgyrtkirk) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java:149]: I wonder if this is new functionality - could it be in a separate PR?

File: sql/src/test/java/org/apache/druid/sql/calcite/parser/DruidSqlUnparseTest.java [hunk 5]
```diff
@@ -58,7 +58,7 @@ public void testUnparseReplaceAll() throws ParseException
                             + "OVERWRITE ALL\n"
                             + "SELECT *\n"
                             + "    FROM \"foo\"\n"
-                            + "PARTITIONED BY ALL TIME "
+                            + "PARTITIONED BY 'ALL TIME' "
```
Comment (kgyrtkirk) [inline sql/src/test/java/org/apache/druid/sql/calcite/parser/DruidSqlUnparseTest.java:61]: is the result of the unparse valid ?

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java (hunk 4 above)
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java]: Thanks! removed this

File: sql/src/test/java/org/apache/druid/sql/calcite/parser/DruidSqlUnparseTest.java (hunk 5 above)
Comment (zachjsh) [reply sql/src/test/java/org/apache/druid/sql/calcite/parser/DruidSqlUnparseTest.java]: Changed back

File: sql/src/main/codegen/includes/common.ftl (hunk 3 above)
Comment (zachjsh) [reply sql/src/main/codegen/includes/common.ftl]: using `SqlLiteral.createSymbol` as you suggested

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java (hunk 1 above)
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:264]: fixed

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/CalcitePlanner.java]: removed
